File: Back End Spring/src/main/java/org/elis/movieexplorer/service/definition/GenereService.java
package org.elis.movieexplorer.service.definition;

import java.util.List;

import org.elis.movieexplorer.dto.genere.request.EditGenereDTO;
import org.elis.movieexplorer.dto.genere.request.InsertGenereDTO;
import org.elis.movieexplorer.dto.genere.response.ResponseGenereDTO;

public interface GenereService {
	// CREATE
	ResponseGenereDTO insert(InsertGenereDTO dto);
	
	// READ
	List<ResponseGenereDTO> findAll();
	
	ResponseGenereDTO findById(Long id);
	
	List<ResponseGenereDTO> findByIdFilm(Long id);
	
	// UPDATE
	ResponseGenereDTO editById(Long id, EditGenereDTO gMod);
	
	// DELETE
	void removeById(Long id);
}

File: Front End Angular/src/app/services/genere.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {InsertGenereDTO} from '../dto/genere/request/insert-genere-dto';
import {finalize, Observable} from 'rxjs';
import {ResponseGenereDTO} from '../dto/genere/response/response-genere-dto';
import {EditGenereDTO} from '../dto/genere/request/edit-genere-dto';
import {SharedService} from "./shared.service";

@Injectable({
  providedIn: 'root'
})
export class GenereService {
  private BASE_URL = 'http://localhost:8080';

  constructor(private http: HttpClient, private sharedService: SharedService) {}


  insert(dto: InsertGenereDTO): Observable<ResponseGenereDTO> {
    return this.http.post<ResponseGenereDTO>(
      `${this.BASE_URL}/staff/genere`,
      dto
    );
  }


  findAll(): Observable<ResponseGenereDTO[]> {
    return this.http.get<ResponseGenereDTO[]>(`${this.BASE_URL}/genere`);
  }


  findById(id: number): Observable<ResponseGenereDTO> {
    this.sharedService.loading = true;

    return this.http.get<ResponseGenereDTO>(`${this.BASE_URL}/genere/${id}`)
        .pipe(finalize(() => this.sharedService.loading = false)
    );
  }


  findByIdFilm(idFilm: number): Observable<ResponseGenereDTO[]> {
    this.sharedService.loading = true;

    return this.http.get<ResponseGenereDTO[]>(
      `${this.BASE_URL}/genere/film/${idFilm}`
    ).pipe(finalize(() => this.sharedService.loading = false));
  }


  editById(id: number, dto: EditGenereDTO): Observable<ResponseGenereDTO> {
    this.sharedService.loading = true;

    return this.http.patch<ResponseGenereDTO>(
      `${this.BASE_URL}/staff/genere/${id}`,
      dto
    ).pipe(finalize(() => this.sharedService.loading = false));
  }


  removeById(id: number): Observable<void> {
    return this.http.delete<void>(`${this.BASE_URL}/staff/genere/${id}`);
  }
}
